intro section first draft (#524)
AdaCore/learn: content/courses/intro-to-embedded-sys-prog/chapters/introduction.rst

Embedded systems don't necessarily involve SoC computers but they 
frequently do. The techniques and information in this course work on any 
of these kinds of computer. 

Down To The Bare Metal
----------------------

Ada has always had facilities designed specifically for embedded 
systems. The language includes constructs for directly manipulating 
hardware, for example, and direct interaction with assembly language. 
These constructs are as effective as those of any high-level programming 
language (yes, including C). These constructs are expressively powerful, 
well-specified (so there are few surprises), efficient, and portable 
(within reason). 

We say "within reason" because portability is a difficult goal for 
embedded systems. That's because the hardware is so much a part of the 
application itself, rather than being abstracted away as in a 
general-purpose application. That said, the hardware details can be 
managed in Ada so that portability is maximized to the extent possible 
for the application. 

Strictly speaking, not all software can or should be absolutely 
portable! If a specific device is required, well, the program won't work 

> with some other device. But tot he extent possible portability is 

Reviewer: ```suggestion
with some other device. But to the extent possible portability is 
```

Author: Got it.

obviously a good thing. 

Separation Principle
--------------------

There is a language design principle underlying the Ada facilities 
intended for implementing embedded software. This design principle 
directly affects how the language is used, and therefore, the 
portability and readability of the resulting application code. 

This language design principle is known as the "separation principle." 
What's being separated? The low-level, less portable aspects of some 
piece of code are separated from the usage of that piece of code. 

Don't confuse this with hiding unnecessary implementation details via 
compile-time visibility control (i.e., information hiding and 
encapsulation). That certainly should be done too. Instead, because of 
the separation principle, we specify the low-level properties of 
something once, when we declare it. From then on, we can use regular Ada 

> code to interact with it. That way the bulk of the code -- the usage -- 

Reviewer: ```suggestion
code to interact with it. That way the bulk of the code |mdash| the usage |mdash| 
```

Author: Got it.
